### PR TITLE
Draft of OSMLR update strategy document.

### PR DESCRIPTION
**Pull request by dnesbitt61**

### CODE REVIEW

File: docs/osmlr_updates.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+Instead, we propose maintaining a single ID "namespace", so that 0/1/1 would always refer to the same segment so long as that segment didn't dramatically change. This requires the OSMLR generation process to begin by matching the previous segment descriptors against the current data, and keeping the IDs for those which continue to match. Since OSMLR descriptors are designed to be tolerant to some kinds of changes, we anticipate that the majority will continue to match. The few which do not match will result in the old IDs being marked as "deleted" or "superseded" and new IDs being allocated for the new segments (possibly with forward / back references to aid in translation).
+
+### Running the OSMLR Update Process
+
+The OSMLR generation process (osmlr) has been updated to accept an optional OSMLR pbf directory. THis indicates that the OSMLR generation process should apply updates based on the supplied OSMLR segment definitions. This triggers the segment "migration" process rather than creating an initial or new set of OSMLR segments.
```
Comment (gknisely) [inline docs/osmlr_updates.md]: THis should be This

File: docs/osmlr_updates.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
...la includes a process to "associate" or match OSMLR Ids and segment definitions to Valhalla edges (or portions of edges). This association can be performed multiple times against an existing OSMLR definition as the underying map data is updated. The association process matches OSMLR segments to roads within specific *tolerances*. These tolerances include how closely the OSMRL Location Reference Points (LRP) match to Valhalla edges and intersections. It also includes tolerances for how closely headings match at LRPs and how closely lengths of the road segment between LRPs match the OSMLR segment definition. Usually the road network data changes slowly (especially on more established, higher classification roads) and the large majority of OSMLR segments properly match the updated Valhalla road graph data. However, some data updates will force existing segments to be deprecated/deleted as the changes result in no matching roads to be found within tolerance of the OSMLR segment. These deleted segments may result in new, replacement OSMLR segments with unique Ids. Also, data updates may represent newly constructed roads (or newly added to the base data) which require new OSMLR segments.
```
Comment (gknisely) [inline docs/osmlr_updates.md]: OSMRL should be OSMLR

File: docs/osmlr_updates.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+* The OSMLR "creation" process is then run to generate new segments that either replace deprecated segments (if applicable) or identify newly added roads within the data. When creating new segments, any road segment that is already associated to an OSMLR is skipped / disallowed. New segments get sequential Ids within the tile, following all currently existing Ids so there is no collision of Ids within the tile. The new segments also have the date indicating when the segment was added to the OSMLR tile.
+
+As the protocol buffer output is updated, the GeoJSON output is similarly updated.
+
+###Common Conditions / Data Changes
+
+THe OSMLR update process tries to maintain existing OSMLR segments, where possible, under minor data edits and changes to the underying roads structure. However, certain changes require replacement of existing OSMLR segments. This section describes some common changes and how OSMLR segments are handled under these conditions.
```
Comment (gknisely) [inline docs/osmlr_updates.md]: THe should be The

File: docs/osmlr_updates.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+* Copy all existing OSMLR segments to the new tile.
+* Identify all OSMLR segments that are associated to the Valhalla routing data. These segments are still valid and remain unchanged with their existing index within the tile remaining unchanged (OSMLR segments are immutable). Skip any previously deprecated OSMLR segments. What remains is the set of OSMLR segments that must be deprecated (they no longer associate to the routing data within specific tolerances). These segments are marked as deprecated/deleted with the date added to specify when the OSMLR segment was deprecated.
+* The OSMLR "creation" process is then run to generate new segments that either replace deprecated segments (if applicable) or identify newly added roads within the data. When creating new segments, any road segment that is already associated to an OSMLR is skipped / disallowed. New segments get sequential Ids within the tile, following all currently existing Ids so there is no collision of Ids within the tile. The new segments also have the date indicating when the segment was added to the OSMLR tile.
+
+As the protocol buffer output is updated, the GeoJSON output is similarly updated.
```
Comment (gknisely) [inline docs/osmlr_updates.md]: As the protocol buffer output is updated, the GeoJSON output is similarly updated by removing the "deleted" or "superseded" features from the featurecollection.  New features are appended to the existing featurecollection for existing GeoJSON tiles.

**Review by gknisely — APPROVED** [review]

**Comment by drewda** [comment]
This is looking good.

Can you list out the tolerances? Or are they set at runtime of the osmlr program?

In the future, diagrams would be useful to describe the "common conditions" at the end -- but not necessary for this first version of docs. Better yet, we'll hopefully be able to capture specific examples when running the update process in the future.

**Comment by dnesbitt61** [comment]
Still working on the tolerances. We have to do some work on tolerances for lengths - likely based on some percentage of the length of the segment (with some minimum). We currently use 20 meters for tolerance of the LRP location and 15 degrees for heading tolerance. But I think we want to do a round of experimentation before nailing down the actual numbers. That is why I was vague in the doc. It might be good to have them be command line (or at least allow overrides of the defaults).

**Comment by drewda** [comment]
O.K. An option is to just list the tolerances and their units in this doc but not give values.

Alternatively, it's possible to link off to a block of code where constants are set or command line arguments are defined.

**Review by drewda — APPROVED** [review]
Looks good to me.

**Pull request merged**